### PR TITLE
Module 7

### PR DESCRIPTION
**Pull request by 8kristy**

### CODE REVIEW

File: .github/workflows/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+name: Continuous Integration
+on: 
+  push:
+    branches: 
+      - main
+    paths-ignore:
+      - "README.md"
+  pull_request:
+    paths-ignore:
+      - "README.md"
```
Comment (8kristy) [inline .github/workflows/pipeline.yml]: I updated the README and pushed and the pipeline still ran on the pull request, did I get something wrong? [GitHub docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-paths) show this as an example
```
on:
  push:
    paths-ignore:
      - 'docs/**'
```

File: .github/workflows/pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
+
+    - name: Send GitHub Action cancelled notification to Slack channel
+      if: steps.slack-secret-check.outputs.available == 'true' && cancelled()
+      uses: slackapi/slack-github-action@v1.26.0
+      with:
+        payload: |
+          {
+            "text": ":warning: GitHub Action cancelled by ${{ github.actor }}"
+          }
```
Comment (8kristy) [inline .github/workflows/pipeline.yml]: Is there a nicer way of doing this without all the repetition, e.g. as a single step with 3 conditions?

File: .github/workflows/build-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    - name: Check if Slack webhook URL secret exists
+      id: slack-secret-check
+      shell: bash
+      if: always()
+      run: |
+        if [ "${{ secrets.SLACK_WEBHOOK_URL }}" != '' ]; then
+          echo "available=true" >> $GITHUB_OUTPUT;
+        else
+          echo "available=false" >> $GITHUB_OUTPUT;
+        fi
```
Comment (8kristy) [inline .github/workflows/build-pipeline.yml:35]: You can't directly access secrets in if statements. I tried using `env` but it didn't quite work... Is there a nicer way of doing this check? This is just something I got of SO

And if not, can I make functions in GitHub actions? I'm using a very similar one for the snyk job, just has different Id and checks different secret

Comment (JackMeadDev) [reply .github/workflows/build-pipeline.yml:35]: Fun question - I had a little play in [this repo](https://github.com/JackMead/github-actions-demo/blob/main/.github/workflows/example.yml) and I think that's meeting your question - we can either determine whether or not to run the job (as we might want to with e.g. Snyk) or just a single step (as with these steps) by converting to environment variables and referencing those:
```
jobs:
    jobConditional:
        env:
            SECRET_EXISTS: ${{ secrets.SECRET_EXISTS }}
        runs-on: ubuntu-latest
        if: ${{ github.env.SECRET_EXISTS != '' }}
        steps:
            - run: echo "This ran"
    stepConditional:
        runs-on: ubuntu-latest
        env:
            SECRET_EXISTS: ${{ secrets.SECRET_EXISTS }}
        steps:
            - run: echo "This ran"
              if: ${{ github.env.SECRET_EXISTS != '' }}
```

File: .github/workflows/security-pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+name: Security
+on:
```
Comment (JackMeadDev) [inline .github/workflows/security-pipeline.yml:2]: Nice work on this one!

**Review by JackMeadDev — APPROVED** [review]
Nice work, I can see your tests are happily running in the pipeline and bonus points for the Snyk & Slack integrations! (And I'm trusting that Slack is working!)😄 I've left a response to the question you posed about the conditional jobs/steps - hopefully I've answered the right question; let me know if not! Happy for you to merge this when you're ready

File: .github/workflows/build-pipeline.yml (hunk 3 above)
Comment (8kristy) [reply .github/workflows/build-pipeline.yml:35]: Think I tried something similar when I said I tried using `env`, tried this suggestions and it still seems to be skipping over so might need to live with the current solution

**Pull request merged**